### PR TITLE
[TACHYON-1247] Fixed deadlock issue in BlockStoreContext and FileSystemContext

### PR DESCRIPTION
**Pull request by yupeng9**
https://tachyon.atlassian.net/browse/TACHYON-1247

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/block/BlockStoreContext.java [hunk 1]
```diff
@@ -108,7 +108,7 @@ public synchronized BlockMasterClient acquireMasterClient() {
    *
    * @param masterClient a block master client to release
    */
-  public synchronized void releaseMasterClient(BlockMasterClient masterClient) {
+  public void releaseMasterClient(BlockMasterClient masterClient) {
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/block/BlockStoreContext.java]: please in javadoc comment why this method (and below) is not synchronized.

**Comment by apc999** [comment]
Also, a high level comment: should we also push this into branch-0.8?

**Comment by aaudiber** [comment]
@yupeng9 Thanks for the fix! Out of curiosity, how did you find the problem?

`ResourcePool` should be thread-safe, so can we drop the corresponding `acquire*Client` `synchonize` modifier?

**Comment by calvinjia** [comment]
@yupeng9 Could you reopen this to branch 0.8 and add a unit test? Also, is this a temporary solution? I'm not sure how this will interact with `resetContext` if it is not synchronized. Perhaps we need a different kind of locking scheme (ie. anything can execute except `resetContext` which should lock everything).

**Comment by yupeng9** [comment]
@aaudiber, I encountered a deadlock when I had several threads concurrently write into Tachyon. @calvinjia, unit tests must be deterministic, so I don't think concurrency can be unit tested in a deterministic way. The closest thing I can think of is to spawn lots of threads, but such tests shouldn't be crafted as unit test. Maybe @jsimsa can shed some light on the best practice on concurrency testing.

**Comment by yupeng9** [comment]
@aaudiber, I think `synchronized` keyword can be dropped on `acquire*Client`. @calvinjia, why do we need `reset` in `*Context`? By looking it use, it's mainly called in initialization and testing, if so, do we need it to be a public method?

**Comment by yupeng9** [comment]
Will reopen in 0.8

**Pull request closed without merging**